src/groupeddataframe/groupeddataframe.jl

```diff
@@ -331,28 +300,48 @@ true
 """
 Base.keys(gd::GroupedDataFrame) = GroupKeys(gd)
 
-# Index with GroupKey
-function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
-    gd === parent(key) && return gd[getfield(key, :idx)]
+
+#
+# Indexing with scalar values
+#
+
+# The allowed key types for dictionary-like indexing
+const GroupKeyTypes = Union{GroupKey, Tuple, NamedTuple}
+# All allowed scalar index types
+const GroupIndexTypes = Union{Integer, GroupKeyTypes}
+
+Compat.lastindex(gd::GroupedDataFrame) = gd.ngroups
+Base.first(gd::GroupedDataFrame) = gd[1]
+Base.last(gd::GroupedDataFrame) = gd[end]
+
+# Single integer indexing (basis for all other scalar methods)
+Base.getindex(gd::GroupedDataFrame, idx::Integer) =
+    view(gd.parent, gd.idx[gd.starts[idx]:gd.ends[idx]], :)
+
+# Find the integer index of a group given any valid scalar key type
+# (Integer case already defined by default, does not accept Bool)
+function Base.to_index(gd::GroupedDataFrame, key::GroupKey)
```

Comment:
Is it really useful to override `Base.to_index` for this? Otherwise, better use a different name to avoid relying on Base internals.

Reply:
It's definitely not necessary, but this is the intended function of `Base.to_index`. I don't think it's actually used by any of the default indexing machinery if the first argument isn't an `AbstractArray`.

Reply:
As a side note: InvertedIndices.jl uses `Base.to_index`.

Reply:
Using `Base.to_indices` in the default `getindex` method, this calls `to_index` when it needs to.